parmed/formats/sdf.py

```diff
@@ -27,10 +27,11 @@ def id_format(filename):
         is_fmt : bool
             True if it is a sdf file, False otherwise
         """
-        line = linecache.getline(filename, lineno=4)
         try:
-            return line.split()[-1] in ['V2000', 'V3000']
+            words = linecache.getline(filename, lineno=4).split()
+            return len(words) >= 3 and words[-1] in ('V2000', 'V3000')
         except IndexError:
+            # 4-th line is empty
             return False
```

Comment:
That's true for any language that short-circuits.  Most languages guarantee short-circuiting in the language standard.  Examples are C (and anything derived from C like C++, Java, Go, C#, etc.), Python, ruby, Swift, perl, and probably awk.

In fact, the only language I know *doesn't* guarantee it in the spec is Fortran.  But the major compilers all implement it, anyway.  What this means is you can do something like this in C:

```c
if (denom != 0 && numerator/denom > 0) {
    // Add some code
}
```

and be guaranteed not to get a divide-by-zero error. But if you tried that in Fortran, the standard does not require compilers to skip the divide-by-zero (although I've used that kind of pattern before and never seen it fail in practice with modern F90 compilers).

Short-circuiting is also responsible for these tricks in Python:

```python
import random

def stupid():
    a = random.random() - 0.5
    mylist = []
    if a > 0:
        mylist.append(0)
    return mylist or 'The list was empty!'

stupid()
```

If you run this a number of times, you'll see output like the following:

```python
>>> stupid()
'The list was empty!'
>>> stupid()
[0]
>>> stupid()
'The list was empty!'
>>> stupid()
'The list was empty!'
>>> stupid()
[0]
```

As a general rule, boolean operations will return the first value that duplicates the total expression's boolean return value.  So if `mylist` is filled with a 0 (which happens ~50% of the time), then `mylist or 'The list was empty!'` evaluates to `True` because `mylist` is boolean `True`.  Short-circuiting allows the boolean value of `mylist` to represent the boolean value of the whole expression.  On occasions where `mylist` does not get appended, the empty list evaluates to `False` and the second value in the expression (the static string `The list was empty!`) represents the boolean value of the expression, so it gets returned instead.

In fact, this even works if the second value is boolean `False` (try replacing `The list was empty` with `False` and you'll see it returns both `[0]` and `False`.  I like this trick and use it on occasion where it simplifies some conditionals.  [See here for example](https://github.com/ParmEd/ParmEd/blob/master/parmed/amber/netcdffiles.py#L777-L778):

```python
    def __del__(self):
        self.closed or (hasattr(self, '_ncfile') and self._ncfile.close())
```

If `self.closed` is True, it satisfies the whole conditional and the second part is skipped.  Otherwise, it moves on to `(hasattr(self, '_ncfile') and self._ncfile.close())`.  If the object died in the constructor, `_ncfile` will not be set as an attribute, so the `hasattr` call will be `False`.  Since it's connected by `and`, the second part of that conditional is short-circuited and the `AttributeError` is avoided.  Therefore, `self._ncfile.close()` is only executed if and only if the object went through its constructor successfully *and* it was not closed explicitly before.  The equivalent code is this:

```python
    def __del__(self):
        if not self.closed and hasattr(self, '_ncfile'):
            self._ncfile.close()
```

But I like my trick, even if I'd be yelled at in the corporate world for using it :smiley:.

One of the main places I use this kind of trick is doing something like this:

```python
return return_value or []
```

to return an empty list instead of `None` in some relevant situations.  Because that replaces something like this:

```python
if return_value is None:
    return []
else:
    return return_value
```

or more succinctly

```python
if return_value is None:
    return []
return return_value
```